### PR TITLE
Foldable block matcher fix

### LINKED ISSUE #131: Foldable block matcher doesn't work properly
From the code below
```java
class MyClass {
  /// Calculates the factorial of the number.
  /// The number must be >= 0.
  static int factorial(int n) {
    if (n == 0) {
      return 1;
    }

    return n * factorial(n - 1);
  }

  public static void main(String[] args) {
    int num = 5;s
    System.out.println("Factorial of " + num + " is " + factorial(5));
  }
}
```
Steps to reproduce:

- Fold at the 4th line
- Move cursor to the end of 13th line
- Enter any character

`Expected`
- character is inserted to the text at cursor location.
- folded block preserve its folding

`Actual`:
- character is inserted at the end of 13th line
- folded block is unfolded
- cursor is moved to the end of 4th line

### PR DESCRIPTION
**Pull request by yescorp**
- Resolves #131

CC https://github.com/apache/beam/issues/25279

### CODE REVIEW

File: test/src/code_field/code_controller_folding_editing_test.dart [hunk 1]
```diff
@@ -146,9 +151,297 @@ int n;
...
+              language: python,
+            );
+
+            controller.foldAt(0);
+
+            expect(controller.value.text, 'a:\n\n');
+            //                                 \ selection
+            await wt.selectFromHome(3);
+
+            controller.value = controller.value.replacedSelection('\na');
```
Comment (alexeyinkin) [inline test/src/code_field/code_controller_folding_editing_test.dart:275]:
```suggestion
            controller.value = controller.value.replacedSelection('\n');
```
It a timer issue arises, debug it.

**Review by alexeyinkin — CHANGES_REQUESTED** [review]

File: lib/src/gutter/gutter.dart [hunk 2]
```diff
@@ -132,6 +132,25 @@ class GutterWidget extends StatelessWidget {
             : () => codeController.foldAt(block.firstLine),
       );
     }
+
+    // Add folded blocks that are not considered as a valid foldable block,
+    // but should be folded because they were folded before becoming invalid.
+    for (final block in code.foldedBlocks) {
+      if (code.lines.length <= block.lastLine) {
```
[the remark this answers is not in the document]
Comment (yescorp) [reply lib/src/gutter/gutter.dart:139]: This is an `early return condition` 
If we change the positions the condition should be the same I think.

File: test/src/code_field/code_controller_folding_editing_test.dart [hunk 3]
```diff
@@ -203,6 +203,77 @@ int n;
             controller.value = controller.value.replacedSelection('');
             expect(controller.value, expected);
           });
+
+          testWidgets('Add new import after folded imports block', (wt) async {
+            const example = 'package example;\nimport java.util.date;\n';
```
Comment (alexeyinkin) [inline test/src/code_field/code_controller_folding_editing_test.dart:208]:
```suggestion
            const example = 'package example;\nimport java.util.Date;\n';
```

File: test/src/code_field/code_controller_folding_editing_test.dart [hunk 4]
```diff
@@ -203,6 +203,77 @@ int n;
             controller.value = controller.value.replacedSelection('');
             expect(controller.value, expected);
           });
+
+          testWidgets('Add new import after folded imports block', (wt) async {
```
Comment (alexeyinkin) [inline test/src/code_field/code_controller_folding_editing_test.dart:207]:
```suggestion
          testWidgets('Adding an import after folded imports -> Unfold', (wt) async {
```

File: test/src/code_field/code_controller_folding_editing_test.dart [hunk 5]
```diff
@@ -203,6 +203,77 @@ int n;
...
+            await wt.cursorEnd();
+            controller.value = controller.value.replacedSelection(inserted);
+            // TODO(yescorp): Fix selection, then fix tests
+            //  https://github.com/akvelon/flutter-code-editor/issues/182
+            expect(controller.value.text, example + inserted);
+            expect(controller.code.foldedBlocks.length, 0);
+          });
+
+          testWidgets(
+              'Add anything but imports line after folded imports block',
```
Comment (alexeyinkin) [inline test/src/code_field/code_controller_folding_editing_test.dart:224]:
```suggestion
              'Adding non-import after folded imports -> Does not unfold',
```

File: test/src/code_field/code_controller_folding_editing_test.dart [hunk 6]
```diff
@@ -272,12 +343,13 @@ int n;
...
 
-            controller.value = controller.value.replacedSelection('\na');
+            controller.value = controller.value.replacedSelection('\n');
 
-            const expectedText = 'a:\n\na\n';
+            const expectedText = 'a:\n\n\n';
 
             expect(controller.value.text, expectedText);
             expect(controller.code.foldedBlocks.length, 1);
+            await wt.moveCursor(-1); // reset timer
```
Comment (alexeyinkin) [inline test/src/code_field/code_controller_folding_editing_test.dart:352]:
```suggestion
            await wt.moveCursor(-1); // Reset timer.
```

**Review by alexeyinkin — CHANGES_REQUESTED** [review]

File: test/src/code_field/code_controller_folding_editing_test.dart [hunk 7]
```diff
@@ -349,18 +350,14 @@ int n;
...
 
       group('Folded block is no longer recognizable', () {
-        group(
-            'Miltiline comment that started before folded block '
-            '1. Doesn\' unfold the block '
-            '2. The block is still unfoldable manually '
-            '3. The block is deletable', () {
-          testWidgets('Braces. 1.', (wt) async {
+        group('Miltiline comment that started before folded block', () {
+          testWidgets('1. Doesn\' unfold the block', (wt) async {
```
Comment (alexeyinkin) [inline test/src/code_field/code_controller_folding_editing_test.dart:360]:
```suggestion
          testWidgets('Does not unfold the block', (wt) async {
```

**Review by alexeyinkin — CHANGES_REQUESTED** [review]

**Review by alexeyinkin — APPROVED** [review]

**Pull request merged**